File: 01/main.clj
(defn toInt [s] 
  (if (empty? s) 0 (Integer. s)))

(def lines (.split (slurp "input.txt") "\n"))

(def calories
  (reverse
    (sort
      (map 
        #(reduce + (map toInt %))
        (partition-by 
          empty?
          lines))))) 

(println "part 1" (first calories))
(println "part 2" (reduce + (take 3 calories)))

;; do a single-pass instead

(def single-pass-calories
  (reverse
    (sort
      (loop
        [total 0 acc [] lines lines]
        (if (empty? lines)
          acc
          (let [n (toInt (first lines))]
            (if (zero? n)
              (recur 0 (conj acc total)(rest lines))
              (recur (+ n total) acc (rest lines)))))))))


(println "part 1 (single-pass)" (first single-pass-calories))
(println "part 2 (single-pass)" (reduce + (take 3 single-pass-calories)))
        

File: 01/main.mjs
import * as fs from 'fs';

const take = (a,n) => a.slice(0,n);
const first = a => a[0];
const last = a => a[a.length-1];
const sum = (a,b) => a+b;
const reverseOrder = (a,b) => a < b ? 1 : (a > b) ? -1 : 0;
const readLines = fileName => fs.readFileSync(fileName, "utf-8").split("\n")

const processLine = ({score,scores}, line)=> {
  if (line) {
    score += Number(line);
  }
  else {
    scores.push(score);
    score = 0;
  }
  return ({score, scores});
};

const scores = readLines("input.txt") 
  .reduce(processLine, {score:0,scores:[]})
  .scores
  .sort(reverseOrder);

console.log("part 1:", first(scores));
console.log("part 2:", take(scores,3).reduce(sum));
